### ISSUE TITLE
Improve page checksum validation.

### ISSUE THREAD

**Issue by dwsteele** [pgbackrest/pgbackrest#952]

**Comment by dwsteele** [comment]
Before working on these issues refactor the page checksum code to work better in C.  Most of the functions were designed to interface with Perl and are no longer needed.  It also makes sense to sync more closely with the PostgreSQL code now that we have an established method for copying code from that project.

* Centralize PostgreSQL page header data structures. (9d488822)

These data structures were copied a few places (but only once in the core code) so put them in a place where everyone can use them.

To do this create a new file, static.auto.h, to contain data types and macros that have stayed the same through all the versions of PostgreSQL that we support.  This allows us to have single, non-versioned set of headers and code for stable data structures like page headers.

Migrate a few types from version.auto.h that are required for page header structures and pull the remaining types from PostgreSQL directly.

We had previously renamed xlog to wal so update those where required since we won't be modifying the PostgreSQL names anymore.

* Remove obsolete function pageChecksumBufferTest(). (a86253f112)

This function made validation faster in Perl because fewer calls (and buffer transformations) were required when all checksums were valid.

In C calling pageChecksumTest() directly is just as efficient so there is no longer a need for pageChecksumBufferTest().

* Rename old page*() functions to conform to new conventions. (91f321fb865)

The general convention now is to prefix PostgreSQL functions with "pg".

* Use PG_PAGE_SIZE_DEFAULT constant instead of pageSize variable. (4ab8943ca886e)

Page size is passed around a lot but in fact it can only have one value, PG_PAGE_SIZE_DEFAULT, which is checked when pg_control is loaded. There may be an argument for supporting multiple page sizes in the future but for now just use the constant to simplify the code.

There is also a significant performance benefit.  Because pageSize was being used in pageChecksumBlock() the main loop was neither unrolled nor vectorized (-funroll-loops -ftree-vectorize) as it is now with a constant loop boundary.

* Remove invalid const in pgPageChecksum() parameter. (77853d3c1)

pgPageChecksum() must modify the page header in order to calculate the checksum.  The modification is temporary but make it clear that it happens by removing the const.

Also make a note about our non-entirely-kosher usage of a const Buffer in the PageChecksum filter.  This is safe as currently coded but at the least we need to be aware of what is going on.

* Use static checksums in mock/all integration tests. (eb4347f20b8)

Using static values serves as a better cross-check against the page checksum code. The downside is that these checksums may not work with some big endian systems but in that case neither will the unit tests.

We can also remove the page checksum interface from LibC which brings us one step closer to eliminating it.

* Remove invalid page checksum test. (1b647a1a22ae36)

All zero pages should not have checksums.  Not only is this test invalid but it will not work with the stock page checksum implementation in PostgreSQL, which checks for zero pages.  Since we will be using that code verbatim soon this test needs to go.

* Use stock PostgreSQL page checksum implementation. (3796b74dcac29)

We were using a customized version which worked fine but was hard to merge with upstream changes.  Now this code is maintained much like the types in static.auto.h that we copy and check with each release.

The goal is to eventually build directly against PostgreSQL (either source or libcommon) and this brings us one step closer.

* Move logic from postgres/pageChecksum to command/backup/pageChecksum(). (e55443c890)

The postgres/pageChecksum module was designed as an interface to the C structs for the Perl code.  The new C code can do this directly so no need for an interface.

Move the remaining test for pgPageChecksum() into the postgres/interface test module.

**Comment by dwsteele** [comment]
Files can grow or shrink during backups which is normal.  However, we would prefer not to copy the portion of files that have grown since the manifest was built because that portion will contain all-new blocks that might be torn and will need to be tested to be sure they are in a valid range for the backup.  So, we will limit reads to the size reported when the manifest was build and skip all the new blocks.

Here are some preparatory commits:

* Allow storage reads to be limited by bytes. (f2548f45ce4c1e444e9)

The current use case is reading files from the PostgreSQL cluster during backup.

A file may grow during backup but we only need to copy the number of bytes that were reported during the manifest build.  The rest will be rebuilt from the WAL during recovery so copying more is just a waste of space.

Limiting the copy sizes in backup will be part of a future commit.

Improve testing of files that change size during the backup. (26c89b2c8c9ea362)

Files can change size during a backup so update and add tests to cover the various scenarios more thoroughly.

**Comment by dwsteele** [comment]
Closing this since the issues that can be fixed have been fixed.
